File: fotoapparat/src/test/java/io/fotoapparat/parameter/selector/PreviewFpsRangeSelectorsTest.java
package io.fotoapparat.parameter.selector;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import io.fotoapparat.parameter.range.Range;
import io.fotoapparat.parameter.range.Ranges;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

public class PreviewFpsRangeSelectorsTest {

    @Test
    public void fromExactFps_Available() throws Exception {
        // Given
        List<Range<Integer>> availableRanges = Arrays.<Range<Integer>>asList(
                Ranges.continuousRange(24000, 24000),
                Ranges.continuousRange(30000, 30000),
                Ranges.continuousRange(40000, 40000)
        );

        // When
        Range<Integer> result = PreviewFpsRangeSelectors
                .fromExactFps(30)
                .select(availableRanges);

        // Then
        assertEquals(
                Ranges.continuousRange(30000, 30000),
                result
        );
    }

    @Test
    public void fromExactFps_NotAvailable() throws Exception {
        // Given
        List<Range<Integer>> availableRanges = Arrays.<Range<Integer>>asList(
                Ranges.continuousRange(24000, 30000),
                Ranges.continuousRange(30000, 36000),
                Ranges.continuousRange(40000, 40000)
        );

        // When
        Range<Integer> result = PreviewFpsRangeSelectors
                .fromExactFps(30)
                .select(availableRanges);

        // Then
        assertNull(result);
    }

    @Test
    public void nearestToExactFps_ExactAvailable() throws Exception {
        // Given
        List<Range<Integer>> availableRanges = Arrays.<Range<Integer>>asList(
                Ranges.continuousRange(24000, 30000),
                Ranges.continuousRange(30000, 30000),
                Ranges.continuousRange(30000, 34000),
                Ranges.continuousRange(30000, 36000)
        );

        // When
        Range<Integer> result = PreviewFpsRangeSelectors
                .nearestToExactFps(30)
                .select(availableRanges);

        // Then
        assertEquals(
                Ranges.continuousRange(30000, 30000),
                result
        );
    }

    @Test
    public void nearestToExactFps_NoExactAvailable() throws Exception {
        // Given
        List<Range<Integer>> availableRanges = Arrays.<Range<Integer>>asList(
                Ranges.continuousRange(24000, 30000),
                Ranges.continuousRange(30000, 34000),
                Ranges.continuousRange(30000, 36000)
        );

        // When
        Range<Integer> result = PreviewFpsRangeSelectors
                .nearestToExactFps(30)
                .select(availableRanges);

        // Then
        assertEquals(
                Ranges.continuousRange(30000, 36000),
                result
        );
    }

    @Test
    public void rangeWithHighestFps() throws Exception {
        // Given
        List<Range<Integer>> availableRanges = Arrays.<Range<Integer>>asList(
                Ranges.continuousRange(24000, 30000),
                Ranges.continuousRange(30000, 34000),
                Ranges.continuousRange(30000, 36000)
        );

        // When
        Range<Integer> result = PreviewFpsRangeSelectors
                .rangeWithHighestFps()
                .select(availableRanges);

        // Then
        assertEquals(
                Ranges.continuousRange(30000, 36000),
                result
        );
    }

    @Test
    public void rangeWithLowestFps() throws Exception {
        // Given
        List<Range<Integer>> availableRanges = Arrays.<Range<Integer>>asList(
                Ranges.continuousRange(24000, 30000),
                Ranges.continuousRange(30000, 34000),
                Ranges.continuousRange(30000, 36000)
        );

        // When
        Range<Integer> result = PreviewFpsRangeSelectors
                .rangeWithLowestFps()
                .select(availableRanges);

        // Then
        assertEquals(
                Ranges.continuousRange(24000, 30000),
                result
        );
    }
}

File: fotoapparat/src/test/java/io/fotoapparat/selector/PreviewFpsRangeSelectorsTest.kt
package io.fotoapparat.selector

import io.fotoapparat.parameter.FpsRange
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class PreviewFpsRangeSelectorsTest {

    @Test
    fun `Select fps range by exact fps value which is available`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 24000),
                FpsRange(30000, 30000),
                FpsRange(40000, 40000)
        )

        // When
        val result = exactFixedFps(30f)(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(30000, 30000),
                actual = result
        )
    }

    @Test
    fun `Select fps range by exact fps value which is not available`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 30000),
                FpsRange(30000, 36000),
                FpsRange(40000, 40000)
        )

        // When
        val result = exactFixedFps(30f)(availableRanges)

        // Then
        assertNull(result)
    }

    @Test
    fun `Select fps range which contains fps and fixed value is available`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 30000),
                FpsRange(30000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000)
        )

        // When
        val result = containsFps(30f)(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(30000, 30000),
                actual = result
        )
    }

    @Test
    fun `Select fps range which contains fps and not-fixed value is available`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000)
        )

        // When
        val result = containsFps(30f)(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(30000, 36000),
                actual = result
        )
    }

    @Test
    fun `Select highest fps`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 24000),
                FpsRange(24000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000),
                FpsRange(36000, 36000)
        )

        // When
        val result = highestFps()(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(30000, 36000),
                actual = result
        )
    }

    @Test
    fun `Select lowest fps`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 24000),
                FpsRange(24000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000),
                FpsRange(36000, 36000)
        )

        // When
        val result = lowestFps()(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(24000, 30000),
                actual = result
        )
    }

    @Test
    fun `Select highest non-fixed fps`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 24000),
                FpsRange(24000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000),
                FpsRange(36000, 36000)
        )

        // When
        val result = highestNonFixedFps()(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(30000, 36000),
                actual = result
        )
    }

    @Test
    fun `Select lowest non-fixed fps`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 24000),
                FpsRange(24000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000),
                FpsRange(36000, 36000)
        )

        // When
        val result = lowestNonFixedFps()(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(24000, 30000),
                actual = result
        )
    }

    @Test
    fun `Select highest fixed fps`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 24000),
                FpsRange(24000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000),
                FpsRange(36000, 36000)
        )

        // When
        val result = highestFixedFps()(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(36000, 36000),
                actual = result
        )
    }

    @Test
    fun `Select lowest fixed fps`() {
        // Given
        val availableRanges = listOf(
                FpsRange(24000, 24000),
                FpsRange(24000, 30000),
                FpsRange(30000, 34000),
                FpsRange(30000, 36000),
                FpsRange(36000, 36000)
        )

        // When
        val result = lowestFixedFps()(availableRanges)

        // Then
        assertEquals(
                expected = FpsRange(24000, 24000),
                actual = result
        )
    }

}